### ISSUE TITLE
Use a _single_ log file on Automation and read stream into Python

### ISSUE THREAD

**Issue by kylebarron** [kylebarron/stata_kernel#54]
@mcaceresb
I was thinking about Mata support #26 , when I realized that Mata wouldn't work on Windows/mac Automation because I export the log file after every command. So when the user sends `mata`, it tries to open a log, run `mata`, and close the log. This doesn't work for obvious reasons.

I think it should be possible to use a _single_ log file that stays open for the _entire_ session.

When I do 
```stata
log using test.log
di 1
```
and `tail -f test.log` in Git bash on Windows, I'm able to read text in a stream as it's written to the file.

As long as my timing between Python and Windows is accurate and I know when Stata is free, (using `UtilIsStataFree`), this should be possible to read into Python as a stream. I've never worked with a streaming text file but it should be fun to test out. Check out: https://docs.python.org/3/library/io.html

![image](https://user-images.githubusercontent.com/15164633/43963853-7facb1a8-9c89-11e8-9aa9-afb7f2b22cc3.png)
![image](https://user-images.githubusercontent.com/15164633/43963862-86466ae0-9c89-11e8-98a9-7fcc99fc51ad.png)

**Comment by mcaceresb** [comment]
You can wrap the call in `stata`, no? `stata("log close test")` is a valid mata call.
![mata_log](https://user-images.githubusercontent.com/4219197/43964656-59c79082-9c8b-11e8-9603-31b0f7be4681.png)

**Comment by mcaceresb** [comment]
Btw \`" "' strings also work in mata, so you can have quoted paths and what not.

**Comment by kylebarron** [comment]
I've never looked into mata but it's good to know that `stata(log close)` would work. 

The much more interesting part of this is that almost certainly I can send results back to Jupyter as a stream instead of once as a string.

Consider
```stata
. foreach i in 1 2 3 4 {
  2. sleep 1000
  3. di `i'
  4. }
```
In the Stata console, each number comes back one at a time. Contrast that with the current nature of the Jupyter kernel. It waits for the entire command to be finished, then reads the log file. It's clear that the stream data is more interesting.

It appears that with pexpect you can [log output as a stream](https://pexpect.readthedocs.io/en/stable/api/pexpect.html#pexpect.spawn.logfile_read) (I have to look more into that)? And obviously if there's a single log file from Automation, that can be read as a stream too.

In any case, I've never worked with streaming files before and I'd be quite excited to learn how to do it. I think this is important, but it could end up changing the control flow of the package somewhat.

**Comment by mcaceresb** [comment]
Yes, redirecting the flow so that the log output is printed as a it comes while waiting for Stata to finish is much better. Stata prints output to its log as it happens, so it would allow long-running commands to display their output as it occurs vs waiting for the entire block to finish, as you say.

However, I'm not 100% sure what you mean by a stream? I thought `open` did the trick:

```python
>>> from os import system
>>> rc = system("touch test.log")
>>> fh = open("test.log", "r")
>>> header = fh.read()
>>> rc = system("echo 1 >> test.log")
>>> fh.read()
'1\n'
>>> rc = system("echo 2 >> test.log")
>>> rc = system("echo 3 >> test.log")
>>> fh.read()
'2\n3\n'
>>> fh.read()
''
>>> fh.close()
```

**Comment by kylebarron** [comment]
Apparently there's a difference between `open` and `io.StringIO()`?
https://docs.python.org/3/library/io.html#text-i-o

I don't know. I've never used it before. 

For reference: https://pymotw.com/3/io/

**Comment by mcaceresb** [comment]
I've often used BytesIO to parse binaries I download, but I've never used StringIO. I imagine it's a similar idea. Bonus points in that it allows you to redirect anything to the stream, not just stuff in the log file.

**Comment by kylebarron** [comment]
> allows you to redirect anything to the stream, not just stuff in the log file.

What does this mean?

**Comment by mcaceresb** [comment]
The link you sent showed that you would write stuff to the stream, print stuff to the stream, and presumably read stuff from a file into a stream. It just seems more versatile is all I meant.

**Comment by kylebarron** [comment]
Fix autocompletions here:
https://github.com/kylebarron/stata_kernel/pull/66#discussion-diff-209467424R229
